Fix part of #5: Abstraction on top of Oppia GAE for ExplorationHandler (Part 2)

rt4914 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
This PR contains everything related to ExplorationHandler, i.e., ExplorationService, Moshi Models, Test cases.
Also, test-cases for ClassroomService have been created.
This PR was earlier on #66 
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is **assigned** to an appropriate reviewer.

Commit:
Classroom testcases and exploration service, models and testcases

veena14cs reviewed (approved):
Looks good.

nikitamarysolomanpvt reviewed (approved):
LGTM

rt4914 commented:
@BenHenning Waiting for your review.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/model/GaeSubtitledHtml.kt:
```diff
@@ -0,0 +1,16 @@
+data class GaeSubtitledHtml(
+
+  @Json(name = "html") val html: String,
+  @Json(name = "content_id") val contentId: String
```
Why not use nullable types for html & content_id per other GAE data classes?

rt4914 replied:
Updated. Added nullable

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/model/GaeVoiceover.kt:
```diff
@@ -0,0 +1,17 @@
+data class GaeVoiceover(
+
+  @Json(name = "file_size_bytes") val fileSizeBytes: Long,
+  @Json(name = "needs_update") val isUpdatedNeeded: Boolean,
```
Nit: ``isUpdateNeeded``

rt4914 replied:
Updated. `isUpdateNeeded`

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/model/GaeVoiceover.kt:
```diff
@@ -0,0 +1,17 @@
+@JsonClass(generateAdapter = true)
+data class GaeVoiceover(
+
+  @Json(name = "file_size_bytes") val fileSizeBytes: Long,
```
Ditto here--why not use nullable types for these three?

rt4914 replied:
Updated. Added nullable

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/model/GaeRecordedVoiceovers.kt:
```diff
@@ -0,0 +1,15 @@
+@JsonClass(generateAdapter = true)
+data class GaeRecordedVoiceovers(
+
+  @Json(name = "voiceovers_mapping") val voiceoversMapping: Map<String, Map<String, GaeVoiceover>>
```
Ditto here--why not nullable?

rt4914 replied:
Updated. Added nullable.

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/model/GaeWrittenTranslation.kt:
```diff
@@ -0,0 +1,16 @@
+@JsonClass(generateAdapter = true)
+data class GaeWrittenTranslation(
+
+  @Json(name = "html") val html: String,
```
Ditto for these two fields--why not nullable?

rt4914 replied:
Updated. Added nullable

BenHenning commented on data/src/main/java/org/oppia/data/backends/gae/model/GaeWrittenTranslations.kt:
```diff
@@ -0,0 +1,15 @@
+@JsonClass(generateAdapter = true)
+data class GaeWrittenTranslations(
+
+  @Json(name = "translations_mapping") val translationsMapping: Map<String, Map<String, GaeWrittenTranslation>>
```
And for this--should it be nullable?

rt4914 replied:
Updated. Added nullable

BenHenning commented on data/src/test/java/org/oppia/data/backends/FakeJsonResponse.kt:
```diff
@@ -13,5 +13,11 @@ object FakeJsonResponse {
+    "{\"is_super_admin\":true,\"topic_summary_dicts\":[{\"version\":3,\"language_code\":\"en\",\"subtopic_count\":0,\"canonical_story_count\":1,\"uncategorized_skill_count\":0,\"topic_model_last_updated\":1567587844908.068,\"id\":\"BpslppK1Tb89\",\"additional_story_count\":0,\"total_skill_count\":0,\"topic_model_created_on\":1567587801448.179,\"name\":\"Math\"}],\"user_email\":\"test@example.com\ ...
+  /** Fake json response for [MockExplorationService] */
+  const val EXPLORATION_SERVICE_RESPONSE: String =
+    "{\"exploration\":{\"param_changes\":[],\"title\":\"What is a Fraction? Part 1\",\"states\":{\"Q6 on N and D\":{\"param_changes\":[],\"interaction\":{\"solution\":null,\"answer_groups\":[{\"outcome\":{\"param_changes\":[],\"feedback\":{\"content_id\":\"feedback_1\",\"html\":\"<p>That's great! You've really learned this well!<\\/p>\"},\"dest\":\"Final Card\",\"refresher_exploration_id\":null,\ ...
```
At this point, it may be worth having a test data folder where we save these JSON files & retrieve them via a loader. It's hard to read this, and I suspect it'll be hard to update as the exploration structure changes.

BenHenning commented on data/src/test/java/org/oppia/data/backends/test/MockClassroomTest.kt:
```diff
@@ -0,0 +1,54 @@
+    val classroomResponse = classroom.execute()
+
+    assertThat(classroomResponse.isSuccessful).isTrue()
+    assertThat("Math").isEqualTo(classroomResponse.body()!!.topic_summary_dicts?.get(0)?.name)
```
I believe this truth line is backwards. It should be: ``assertThat(testResult).isEqualTo(expectedOutcome)``

rt4914 replied:
Updated all test cases accordingly.

BenHenning commented on data/src/test/java/org/oppia/data/backends/FakeJsonResponse.kt:
```diff
@@ -13,5 +13,11 @@ object FakeJsonResponse {
+  const val CLASSROOM_SERVICE_RESPONSE: String =
+    "{\"is_super_admin\":true,\"topic_summary_dicts\":[{\"version\":3,\"language_code\":\"en\",\"subtopic_count\":0,\"canonical_story_count\":1,\"uncategorized_skill_count\":0,\"topic_model_last_updated\":1567587844908.068,\"id\":\"BpslppK1Tb89\",\"additional_story_count\":0,\"total_skill_count\":0,\"topic_model_created_on\":1567587801448.179,\"name\":\"Math\"}],\"user_email\":\"test@example.com\ ...
+  /** Fake json response for [MockExplorationService] */
+  const val EXPLORATION_SERVICE_RESPONSE: String =
```
We should use one of the demo explorations instead: https://github.com/oppia/oppia/tree/develop/data/explorations since these are already checked into the Oppia repository to be used for this purpose.

rt4914 replied:
I tried converting these YAML files to JSON, but online converters are giving error that the YAML is invalid.

seanlip replied:
Hi @rt4914 -- in general, when reporting an error / invalidity, please could you also explain _why_ it's invalid? That will help with debugging/resolving the issue, otherwise the discussion stalls. (This is important to do in future discussions, too.)

Thanks!

BenHenning reviewed (changes requested):
Generally LGTM. If possible, I'd like to do another pass today before approving.

Commit:
Load json files for test cases

Commit:
Remove unused imports

rt4914 commented:
@BenHenning PTAL. I have introduced json files in `test` folders. Removed `FakeJsonResponse` and created an `ApiUtils` file which will load data from json

rt4914 commented:
https://github.com/oppia/oppia-android/pull/99#discussion_r324040078

@seanlip Hi sean, actually the discussion shifted to hangout chats (between me, veena, nikita and Ben) with details and because of which it missed on here. 

For reference: Online converters (like http://convertjson.com/yaml-to-json.htm) cannot convert the examples provided in Oppia-Web (https://github.com/oppia/oppia/tree/develop/data/explorations) because of rich-text. Rich-text contains HTML tags which makes the YAML response invalid for the online converters.

Solution: Run local development server -> load explorations -> `http://localhost:8181/explorehandler/init/<exploration_id>` this will provide the json (with XSSI prefix).

rt4914 commented:
@BenHenning project structure for your reference:

<img width="427" alt="Screenshot 2019-09-13 at 12 27 03 PM" src="https://user-images.githubusercontent.com/9396084/64843132-e540ba00-d621-11e9-83af-8471a54b9428.png">

BenHenning reviewed (approved):
Thanks @rt4914! LGTM.
